New evaluation page

mmircea16 opened the pull request:
Should close https://github.com/code4romania/stam-acasa/issues/124

It updates with the new interactions as see in the issue.  Connected to the API to retrieve profile, dependants, form and send form results.

Missing:
- Better styling for personal data in intro for self evaluation
- Connecting with the API that returns the form structure (when ready)
- Maybe changing the styling to look more like figma (especially color-wise) - not sure is needed though

Needs feedback on interactions around questions in the form - might be that some interactions with the form need to updated.

mmircea16 committed:
Fix user management

When navigating from the browser to an AuthenticatedRoute the app acted as being not logged in

mmircea16 committed:
Removing confusing use of a cached user

Holding the user as a cache can cause confusion on how to use the service. UserManager should be in charge of the user

mmircea16 committed:
Added basic tabs for different kinds of evaluation

mmircea16 committed:
Added intro to self evaluation

Also, introduced jest and enzyme for testing

mmircea16 committed:
Extracted an API object to abstract the api calls from react

mmircea16 committed:
Added intro for evaluation for others

mmircea16 committed:
Fixing issue with authentication

mmircea16 committed:
Wiring up the family endpoint and added test data into the db

mmircea16 committed:
Saving dependants form to the database

mmircea16 committed:
Wiring up the profile api plus very skinny profile

mmircea16 committed:
Extracting the API url to an env

mmircea16 committed:
Updating the forms to be as much as in the sketches

mmircea16 committed:
Updating the link in the top bar

mmircea16 committed:
Replacing last 5001 reference

mmircea16 committed:
Merge branch 'develop' into new-evaluation-page

mmircea16 committed:
Refatoring api wrappers

mmircea16 committed:
Added list of dependants to the Self evaluation page

mmircea16 committed:
Preparing for an API call to read the form

mmircea16 committed:
Styling + fix around missing deps on useEffect

mmircea16 committed:
Styling for tabs

mmircea16 committed:
Added go back to main page button at the end of the form

mmircea16 committed:
Added a more gracious handling of the potential API failures + edge case with no dependants

Removed tests as shallow and useEffect are not working together

mmircea16 committed:
Replacing enzyme with react-test-library in the hope it handles useEffect better

mmircea16 committed:
Merge branch 'develop' into new-evaluation-page

mmircea16 committed:
Added back accidentally removed line

tudorgergely commented on frontend/src/components/Evaluation/finishFormButton.js:
```diff
@@ -0,0 +1,23 @@
+  );
+};
+
+export default withRouter(FinishFormButton);
```
use react-router hooks instead (useHistory, useLocation etc)

mmircea16 replied:
Good point! will update

tudorgergely commented on frontend/src/components/Evaluation/tabTitle.js:
```diff
@@ -0,0 +1,28 @@
+  );
+};
+
+export default withRouter(TabTitle);
```
use react hooks from react-router instead

tudorgergely commented on frontend/src/components/OtherEvaluation/index.js:
```diff
@@ -0,0 +1,54 @@
+  const [started, setStarted] = useState(false);
+  const [finished, setFinished] = useState(false);
+  const [dependant, setDependant] = useState(null);
+  const [evaluationForm, setEvaluationForm] = useState(null);
```
we could move this to redux (not store it locally here)

tudorgergely commented on frontend/src/components/OtherEvaluation/introOtherEvaluation.js:
```diff
@@ -0,0 +1,56 @@
+const IntroOtherEvaluation = ({ onFinish }) => {
+  useEffect(() => {
+    ProfileApi.getDependants().then(members => {
+      setDependants(members);
```
move this into redux as well (I've seen it used in another component as well)

tudorgergely replied:
move `useState` calls before this effect since you are using the setters here.

tudorgergely commented on frontend/src/components/OtherEvaluation/introOtherEvaluation.js:
```diff
@@ -0,0 +1,56 @@
+    return {
+      text: `${dependant.name} ${dependant.surname}`,
+      value: dependant.id,
+      selected: dependant.id === selectedMember
```
Is this from our taskforce component? because using selected on options is not recommended (also throws an error in the console because of this) @surdu

mmircea16 replied:
you're right! I looked at storybook and shows `selected`. Will check the code to see how to do it

tudorgergely commented on frontend/src/components/OtherEvaluation/introOtherEvaluation.js:
```diff
@@ -0,0 +1,56 @@
+    return <div>Nu aveti dependenti pentru care sa completati formulare</div>;
+  }
+
+  const options = dependants.map(dependant => {
```
if you move dependants into redux you could use a selector for this

tudorgergely commented on frontend/src/components/SelfEvaluation/index.js:
```diff
@@ -1,25 +1,53 @@
+  const [evaluationForm, setEvaluationForm] = useState(null);
+
+  useEffect(() => {
+    EvaluationApi.getSelfEvaluationForm().then(setEvaluationForm);
```
again, we could move this to redux

tudorgergely commented on frontend/src/components/SelfEvaluation/index.js:
```diff
@@ -1,25 +1,53 @@
+import EvaluationApi from "../../api/evaluationApi";
+import FinishFormButton from "../Evaluation/finishFormButton";
 
 const SelfEvaluation = () => {
```
This component and otherEvaluation are very similar - can we merge them somehow?

mmircea16 replied:
I was wondering about. Stopped short of it - will give it a try to see how readable it it

tudorgergely commented on frontend/src/components/SelfEvaluation/introSelfEvaluation.js:
```diff
@@ -0,0 +1,29 @@
+  }, []);
+  return (
+    <div>
+      <StartFormButton onClick={() => onFinish()} />
```
use onClick={onFinish}

tudorgergely commented on frontend/src/components/SelfEvaluation/profileSummary.js:
```diff
@@ -0,0 +1,60 @@
+
+// eslint-disable-next-line react/prop-types
+const ProfileSummary = ({ profile, dependants }) => {
+  const personalData = {
```
can you extract this into a function(profile)?

mmircea16 committed:
Using hooks for router

mmircea16 committed:
Fix use of apiUrl

mmircea16 committed:
Fixing use of select - it needs to be updated in the components library to remove the browser warnings

mmircea16 committed:
Small refactor - implementing PR feedback

mmircea16 committed:
Extracting common evaluation form

mmircea16 committed:
Fixing use of select - it needs to be updated in the components library

mmircea16 committed:
Fixing browser warnings

Changed behaviour of the tabs to use a link to be more inline with how bulma tabs are should work

tudorgergely commented on frontend/src/components/Evaluation/evaluationForm.js:
```diff
@@ -0,0 +1,71 @@
+
+const EvaluationForm = ({ getForm, sendResults, introType }) => {
+  const [started, setStarted] = useState(false);
+  const finished = useState(false);
```
```suggestion
  const [finished] = useState(false);
```

tudorgergely commented on frontend/src/components/Evaluation/tabTitle.js:
```diff
@@ -0,0 +1,24 @@
+
+const TabTitle = ({ link, children }) => {
+  const history = useHistory();
+  const pathname = useLocation().pathname;
```
```suggestion
  const {pathname} = useLocation();
```

tudorgergely commented on frontend/src/components/Evaluation/tabTitle.js:
```diff
@@ -0,0 +1,24 @@
+  return (
+    <li
+      className={link === pathname && "is-active"}
+      onClick={() => history.push(link)}
```
why don't you use NavLink from react-router?

mmircea16 replied:
I ended up using the `a` tag with `href` as the browser was complaining about the lack of `href`. And bulma needs an `a` tag for styling of tabs

tudorgergely commented on frontend/src/components/OtherEvaluation/introOtherEvaluation.js:
```diff
@@ -0,0 +1,56 @@
+  const [selectedMember, setSelectedMember] = useState(undefined);
+
+  const props = {
+    onChange: function(el) {
+      setSelectedMember(el.target.value);
+    },
```
```suggestion
    onChange: el => setSelectedMember(el.target.value),
```

tudorgergely commented on frontend/src/components/OtherEvaluation/introOtherEvaluation.js:
```diff
@@ -0,0 +1,56 @@
+  useEffect(() => {
+    ProfileApi.getDependants().then(members => {
+      setDependants(members);
+      setSelectedMember(members[0].id);
```
check for empty/null

mmircea16 committed:
Update frontend/src/components/Evaluation/evaluationForm.js

Co-Authored-By: Tudor Gergely <tudorgergely@gmail.com>

mmircea16 committed:
Update frontend/src/components/Evaluation/tabTitle.js

Co-Authored-By: Tudor Gergely <tudorgergely@gmail.com>

mmircea16 committed:
Update frontend/src/components/OtherEvaluation/introOtherEvaluation.js

Co-Authored-By: Tudor Gergely <tudorgergely@gmail.com>

mmircea16 committed:
PR feedback

mmircea16 committed:
Merge branch 'develop' into new-evaluation-page

mmircea16 committed:
Re-enabling finished button at the end of forms
